### PR TITLE
Add all classes

### PR DESCRIPTION
**Pull request by KomaritsaKseniia06**

### CODE REVIEW

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        ArticleCategorySaver saver = context.getBean(ArticleCategorySaver.class);
+        UserNotifier userNotifier = context.getBean(UserNotifier.class);
+
+        analyzer.analyzeArticleContent("articles.json", 0);
+        analyzer.analyzeArticleContent("articles.json", 1);
+        analyzer.analyzeArticleContent("articles.json", 2);
+
+        saver.JsonSaver();
+
+        userNotifier.notifyUsersAboutCategoryChanges();
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java]: move all this business logic into some class, here only should be
```suggestion
analyzer.analyze()
```

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    private final String[] excludedWordsArray;
+    private final List<Set<String>> allCategories;
+
+    public ArticleCategoryAnalyzer(@Value("${excluded.words}") String excludedWords, JsonFileService fileService) {
+        this.excludedWordsArray = excludedWords.split(", ");
+        this.fileService = fileService;
+        this.allCategories = new ArrayList<>(); // Ініціалізуємо список категорій
+    }
+
+    public void analyzeArticleContent(String filePath, int articleId) throws IOException {
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: this method should return results of all found categories from all articles

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        String content = fileService.readJsonFile(filePath, Article.class).get(articleId).getContent();
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: get all articles and analyze them at once, not by id, imagine you don't know how many of them is in the file but still need to get the result

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+        for (String word : wordsArray) {
+            word = word.toLowerCase();
+            if (!isExcludedWord(word)) {
+                wordCounts.put(word, wordCounts.getOrDefault(word, 0) + 1);
+            }
+        }
+
+
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: leave only one empty line

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            if (excludedWord.equalsIgnoreCase(word)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public List<Set<String>> getAllCategories() {
+        return allCategories;
+    }
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: don't use global variables that are mutable, remove allCategories field and this getter, the analyzeArticleContent method should return this field

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+@Service
+public class JsonFileService {
+
+    public <T> List<T> readJsonFile(String filePath, Class<T> classOfT) {
+        try {
+            Resource resource = new ClassPathResource(filePath); // Convert filePath to a string
+            InputStreamReader reader = new InputStreamReader(resource.getInputStream());
+            Type type = TypeToken.getParameterized(List.class, classOfT).getType();
+            return new Gson().fromJson(reader, type);
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: nice method, but do not create every time Gson instance, make it as a bean (in config clas) and inject it here

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public <T> void writeJsonFile(List<Set<String>> list) {
+        Gson gson = new Gson();
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: same here for GSON

File: article-analyzer/src/main/java/com/reader/article_analyzer/Model/User.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        return slackId;
+    }
+
+    public void setSlackId(String slackId) {
+        this.slackId = slackId;
+    }
+
+    private String email;
+    private String slackId;
+    private NotificationType notificationType;
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/Model/User.java]: keep all fields together and all methods together, do not mix them

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+@Component
+public class ArticleCategoryAnalyzer {
+    private final JsonFileService fileService;
+    private final String[] excludedWordsArray;
+    private final List<Set<String>> allCategories;
+
+    public ArticleCategoryAnalyzer(@Value("${excluded.words}") String excludedWords, JsonFileService fileService) {
+        this.excludedWordsArray = excludedWords.split(", ");
+        this.fileService = fileService;
+        this.allCategories = new ArrayList<>(); // Ініціалізуємо список категорій
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: this comment tells the same that the code says to us, remove it

image

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Set;
+
+@Service
+public class JsonFileService {
+
+    public <T> List<T> readJsonFile(String filePath, Class<T> classOfT) {
+        try {
+            Resource resource = new ClassPathResource(filePath); // Convert filePath to a string
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: this comment tells not the same what the code does...

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java (hunk 1 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java]:
```suggestion       

 Analyzer analyzer = context.getBean(Analyzer.class);
 analyzer.analyze();

```

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java:26]: In class ArticleAnalyzerApplication

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java:26]:
```suggestion

import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
public class Analyzer {
    private final ArticleCategoryAnalyzer articleCategoryAnalyzer;
    private final ArticleCategorySaver articleCategorySaver;
    private final UserNotifier userNotifier;


    public Analyzer(ArticleCategoryAnalyzer articleCategoryAnalyzer, ArticleCategorySaver articleCategorySaver, UserNotifier userNotifier) {
        this.articleCategoryAnalyzer = articleCategoryAnalyzer;
        this.articleCategorySaver = articleCategorySaver;
        this.userNotifier = userNotifier;
    }

    public void analyze() throws IOException {
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 0);
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 1);
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 2);
        articleCategorySaver.JsonSaver();
        userNotifier.notifyUsersAboutCategoryChanges();
    }
}

```

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java:26]: import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
public class Analyzer {
    private final ArticleCategoryAnalyzer articleCategoryAnalyzer;
    private final ArticleCategorySaver articleCategorySaver;
    private final UserNotifier userNotifier;


    public Analyzer(ArticleCategoryAnalyzer articleCategoryAnalyzer, ArticleCategorySaver articleCategorySaver, UserNotifier userNotifier) {
        this.articleCategoryAnalyzer = articleCategoryAnalyzer;
        this.articleCategorySaver = articleCategorySaver;
        this.userNotifier = userNotifier;
    }

    public void analyze() throws IOException {
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 0);
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 1);
        articleCategoryAnalyzer.analyzeArticleContent("articles.json", 2);
        articleCategorySaver.JsonSaver();
        userNotifier.notifyUsersAboutCategoryChanges();
    }
}

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleAnalyzerApplication.java]: I changed it and created a new class Analyzer with business logic

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java (hunk 2 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]:
```suggestion
    public List<Set<String>> analyzeArticleContent(String filePath, int articleId) throws IOException {
```

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: Changed the return type

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java (hunk 3 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: Fixed it

Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java:23]: Fixed it

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java (hunk 4 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: Did it

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java (hunk 5 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: Now the analyzeArticleContent method returns field allCategories. However, I need a getter for:
private final List<Set<String>> allCategories;
to retrieve categories in class ArticleCategorySaver

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java (hunk 6 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: Changed it with the creation new class AnalyzerConfiguration

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java (hunk 7 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: Changed it
```suggestion
        String gsonString = gson.toJson(list);
```

File: article-analyzer/src/main/java/com/reader/article_analyzer/Model/User.java (hunk 8 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/Model/User.java]: Got it. It was an accident

File: article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java (hunk 10 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/JsonFileService.java]: Changed it

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java (hunk 9 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategoryAnalyzer.java]: Changed it

File: article-analyzer/src/main/java/com/Analyzer [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+    public Analyzer(ArticleCategoryAnalyzer articleCategoryAnalyzer, ArticleCategorySaver articleCategorySaver, UserNotifier userNotifier) {
+        this.articleCategoryAnalyzer = articleCategoryAnalyzer;
+        this.articleCategorySaver = articleCategorySaver;
+        this.userNotifier = userNotifier;
+    }
+
+    public void analyze() throws IOException {
+        articleCategoryAnalyzer.analyzeArticleContent("articles.json");
+        articleCategorySaver.saveCategories();
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/Analyzer]:
```suggestion
        Set<String> categories = articleCategoryAnalyzer.analyzeArticleContent("articles.json");
        articleCategorySaver.saveCategories(categories);
```

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategorySaver.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    private final JsonFileService jsonFileService;
+
+    public ArticleCategorySaver(ArticleCategoryAnalyzer articleCategoryAnalyzer, JsonFileService jsonFileService) {
+        this.articleCategoryAnalyzer = articleCategoryAnalyzer;
+        this.jsonFileService = jsonFileService;
+    }
+
+    public void saveCategories() throws IOException {
+        jsonFileService.writeJsonFile(articleCategoryAnalyzer.analyzeArticleContent("articles.json").stream().toList());
+    }
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategorySaver.java]:
```suggestion
    public void saveCategories(ІSet<String> categories) throws IOException {
        jsonFileService.writeJsonFile(categories);
    }
```

**Comment by vitaliiStefanchakInventorsoft** [comment]
ConsiderItDoneItsDoneGIF

File: article-analyzer/src/main/java/com/Analyzer (hunk 11 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/Analyzer]: Did it

File: article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategorySaver.java (hunk 12 above)
Comment (KomaritsaKseniia06) [reply article-analyzer/src/main/java/com/reader/article_analyzer/ArticleCategorySaver.java]: Did it

**Pull request closed without merging**